token/program-2022/src/extension/mod.rs

```diff
@@ -836,24 +755,135 @@ impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
                 }
             }
-            self.account_type[0] = S::ACCOUNT_TYPE.into();
+            account_type[0] = S::ACCOUNT_TYPE.into();
         }
         Ok(())
     }
 }
+
+/// Encapsulates mutable base state data (mint or account) with possible
+/// extensions
+#[derive(Debug, PartialEq)]
+pub struct StateWithExtensionsMut<'data, S: BaseState> {
+    /// Unpacked base data
+    pub base: S,
+    /// Raw base data
+    base_data: &'data mut [u8],
+    /// Writable account type
+    account_type: &'data mut [u8],
+    /// Slice of data containing all TLV data, deserialized on demand
+    tlv_data: &'data mut [u8],
+}
+impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
+    /// Unpack base state, leaving the extension data as a mutable slice
+    ///
+    /// Fails if the base state is not initialized.
+    pub fn unpack(input: &'data mut [u8]) -> Result<Self, ProgramError> {
+        check_min_len_and_not_multisig(input, S::LEN)?;
+        let (base_data, rest) = input.split_at_mut(S::LEN);
+        let base = S::unpack(base_data)?;
+        let (account_type, tlv_data) = unpack_type_and_tlv_data::<S>(rest)?;
+        Ok(Self {
+            base,
+            base_data,
+            account_type,
+            tlv_data,
+        })
+    }
+
+    /// Unpack an uninitialized base state, leaving the extension data as a
+    /// mutable slice
+    ///
+    /// Fails if the base state has already been initialized.
+    pub fn unpack_uninitialized(input: &'data mut [u8]) -> Result<Self, ProgramError> {
+        check_min_len_and_not_multisig(input, S::LEN)?;
+        let (base_data, rest) = input.split_at_mut(S::LEN);
+        let base = S::unpack_unchecked(base_data)?;
+        if base.is_initialized() {
+            return Err(TokenError::AlreadyInUse.into());
+        }
+        let (account_type, tlv_data) = unpack_uninitialized_type_and_tlv_data::<S>(rest)?;
+        let state = Self {
+            base,
+            base_data,
+            account_type,
+            tlv_data,
+        };
+        if let Some(extension_type) = state.get_first_extension_type()? {
+            let account_type = extension_type.get_account_type();
+            if account_type != S::ACCOUNT_TYPE {
+                return Err(TokenError::ExtensionBaseMismatch.into());
+            }
+        }
+        Ok(state)
+    }
+
+    /// Packs base state data into the base data portion
+    pub fn pack_base(&mut self) {
+        S::pack_into_slice(&self.base, self.base_data);
+    }
+}
 impl<'a, S: BaseState> BaseStateWithExtensions<S> for StateWithExtensionsMut<'a, S> {
     fn get_tlv_data(&self) -> &[u8] {
         self.tlv_data
     }
 }
+impl<'a, S: BaseState> BaseStateWithExtensionsMut<S> for StateWithExtensionsMut<'a, S> {
+    fn get_tlv_data_mut(&mut self) -> &mut [u8] {
+        self.tlv_data
+    }
+    fn get_account_type_mut(&mut self) -> &mut [u8] {
+        self.account_type
+    }
+}
+
+fn unpack_type_and_tlv_data<S: BaseState>(
```

Comment:
It would be rather nice to dedupe this and the following fn. Could have a common method that takes a "check" closure? Something like:
```
fn generic_unpack_type_and_tlv_data_but_with_better_name<S: BaseState, F>(
    rest: &mut [u8],
    check_fn: F,
) -> Result<(&mut [u8], &mut [u8]), ProgramError>
where F: Fn(AccountType) -> Result<(), ProgramError>
 {
        if let Some((account_type_index, tlv_start_index)) = type_and_tlv_indices::<S>(rest)? {
        // type_and_tlv_indices() checks that returned indexes are within range
        let account_type = AccountType::try_from(rest[account_type_index])
            .map_err(|_| ProgramError::InvalidAccountData)?;
        check_fn(account_type)?;
        let (account_type, tlv_data) = rest.split_at_mut(tlv_start_index);
        Ok((
            &mut account_type[account_type_index..tlv_start_index],
            tlv_data,
        ))
    } else {
        Ok((&mut [], &mut []))
    }
}

fn unpack_type_and_tlv_data<S: BaseState>(
    rest: &mut [u8],
) -> Result<(&mut [u8], &mut [u8]), ProgramError> {
    generic_unpack_type_and_tlv_data_but_with_better_name(rest, check_account_type)
}


fn unpack_uninitialized_type_and_tlv_data<S: BaseState>(
    rest: &mut [u8],
) -> Result<(&mut [u8], &mut [u8]), ProgramError> {
    generic_unpack_type_and_tlv_data_but_with_better_name(
        rest,
        |account_type: AccountType| -> Result<(), ProgramError> {
            if account_type != AccountType::Uninitialized {
                Err(ProgramError::InvalidAccountData)
            } else {
                Ok(())
            }
    })
}
```
(I'm sure that code is missing some `S` generic handling.)

Reply:
Works for me, thanks!

Reply:
I went with the lowest-hanging fruit name of `unpack_type_and_tlv_data_with_check`, let me know how that looks